File: frontend/src/shared/lib/permissions/visibilityScope.ts
import type { VisibilityScope } from '../../api/testObjects';

/** Пустой scope = доступ ко всем лабораториям и подразделениям. */
export function isScopeUnrestricted(
  scope: Pick<VisibilityScope, 'laboratory_ids' | 'department_ids'> | null | undefined
): boolean {
  if (!scope) {
    return true;
  }
  return (scope.laboratory_ids?.length || 0) === 0 && (scope.department_ids?.length || 0) === 0;
}

/**
 * Доступ к лаборатории/подразделению по visibility_scope (как на бэкенде).
 * Пустой scope — везде; department_ids — конкретные подразделения; laboratory_ids — вся лаборатория.
 */
export function isVisibleInScope(
  scope: Pick<VisibilityScope, 'laboratory_ids' | 'department_ids'> | null | undefined,
  laboratoryId?: number | null,
  departmentId?: number | null
): boolean {
  if (isScopeUnrestricted(scope)) {
    return true;
  }

  const laboratoryIds = scope?.laboratory_ids || [];
  const departmentIds = scope?.department_ids || [];

  if (departmentId != null && departmentIds.includes(departmentId)) {
    return true;
  }

  if (laboratoryId != null && laboratoryIds.includes(laboratoryId)) {
    return true;
  }

  return false;
}

/**
 * Карточка лаборатории кликабельна, если лаборатория в scope
 * или в scope есть подразделения.
 */
export function isLaboratoryCardAccessible(
  scope: Pick<VisibilityScope, 'laboratory_ids' | 'department_ids'> | null | undefined,
  laboratoryId: number
): boolean {
  if (isScopeUnrestricted(scope)) {
    return true;
  }

  const laboratoryIds = scope?.laboratory_ids || [];
  const departmentIds = scope?.department_ids || [];

  if (laboratoryIds.includes(laboratoryId)) {
    return true;
  }

  return departmentIds.length > 0;
}

File: backend/utils/visibility_scope.py
from __future__ import annotations
from typing import Any
from schemas.visibility import ScopeIdFilter


def normalize_visibility_scope(
    scope: ScopeIdFilter | dict[str, Any] | None,
) -> ScopeIdFilter:
    """Привести область видимости к виду со списками id лабораторий и подразделений."""
    if not scope or not isinstance(scope, dict):
        return {"laboratory_ids": [], "department_ids": []}

    laboratory_ids = scope.get("laboratory_ids") or []
    department_ids = scope.get("department_ids") or []

    return {
        "laboratory_ids": [int(item) for item in laboratory_ids if isinstance(item, int) and item > 0],
        "department_ids": [int(item) for item in department_ids if isinstance(item, int) and item > 0],
    }


def is_visible_in_scope(
    visibility_scope: ScopeIdFilter | dict[str, Any] | None,
    laboratory_id: int | None = None,
    department_id: int | None = None,
) -> bool:
    """
    Пустая область видимости означает доступность везде.
    Подразделение в department_ids — видно только для этого подразделения.
    Лаборатория в laboratory_ids — видна для всей лаборатории (без подразделений).
    """
    scope = normalize_visibility_scope(visibility_scope)
    laboratory_ids = scope["laboratory_ids"]
    department_ids = scope["department_ids"]

    if not laboratory_ids and not department_ids:
        return True

    if department_id and department_id in department_ids:
        return True

    return bool(laboratory_id and laboratory_id in laboratory_ids)


__all__ = [
    "is_visible_in_scope",
    "normalize_visibility_scope",
]
